pkg/scheduler/framework/plugins/podtopologyspread/pod_topology_spread.go

```diff
@@ -51,7 +110,172 @@ func (pl *PodTopologySpread) Filter(ctx context.Context, cycleState *framework.C
+		// (2) All topologyKeys need to be present in `node`
+		if !predicates.PodMatchesNodeSelectorAndAffinityTerms(pod, node) ||
+			!predicates.NodeLabelsMatchSpreadConstraints(node.Labels, constraints) {
+			return
+		}
+
+		for _, constraint := range constraints {
+			pair := topologyPair{key: constraint.TopologyKey, value: node.Labels[constraint.TopologyKey]}
+			// If current topology pair is not associated with any candidate node,
+			// continue to avoid unnecessary calculation.
+			if pl.podTopologySpreadMap.topologyPairToPodCounts[pair] == nil {
+				continue
+			}
+
+			// <matchSum> indicates how many pods (on current node) match the <constraint>.
+			matchSum := int64(0)
+			for _, existingPod := range nodeInfo.Pods() {
+				match, err := predicates.PodMatchesSpreadConstraint(existingPod.Labels, constraint)
+				if err != nil {
+					errCh.SendErrorWithCancel(err, cancel)
+					return
+				}
+				if match {
+					matchSum++
+				}
+			}
+			atomic.AddInt64(pl.podTopologySpreadMap.topologyPairToPodCounts[pair], matchSum)
+		}
+	}
+	workqueue.ParallelizeUntil(ctx, 16, len(allNodeNames), processAllNode)
+	if err := errCh.ReceiveError(); err != nil {
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+
+	return nil
+}
+
+// Initialized returns if the plugin is initialized.
+func (pl *PodTopologySpread) Initialized() bool {
+	return pl.podTopologySpreadMap != nil
+}
+
+// Score invoked at the Score extension point.
+// The "score" returned in this function is the matching number of pods on the `nodeName`,
+// it needs to normalized later.
+func (pl *PodTopologySpread) Score(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) (int64, *framework.Status) {
+	if !pl.Initialized() {
+		return 0, nil
+	}
+	// no need to continue if the node is not qualified.
+	if _, ok := pl.podTopologySpreadMap.nodeNameSet[nodeName]; !ok {
+		return 0, nil
+	}
+
+	constraints := getSoftTopologySpreadConstraints(pod)
+	nodeInfo, exist := pl.handle.NodeInfoSnapshot().NodeInfoMap[nodeName]
+	node := nodeInfo.Node()
+	if !exist || node == nil {
+		return 0, framework.NewStatus(framework.Error, fmt.Sprintf("node %q does not exist in NodeInfoSnapshot", nodeName))
+	}
+
+	// For each present <pair>, current node gets a credit of <matchSum>.
+	// And we sum up <matchSum> and return it as this node's score.
+	var score int64
+	for _, constraint := range constraints {
+		if tpVal, ok := node.Labels[constraint.TopologyKey]; ok {
+			pair := topologyPair{key: constraint.TopologyKey, value: tpVal}
+			matchSum := *pl.podTopologySpreadMap.topologyPairToPodCounts[pair]
+			score += matchSum
+		}
+	}
+
+	// If needed, modify the shared pl.podTopologySpreadMap.minCount atomically.
+	for oldMinCount := atomic.LoadInt64(&pl.podTopologySpreadMap.minCount); score < oldMinCount; {
+		if atomic.CompareAndSwapInt64(&pl.podTopologySpreadMap.minCount, oldMinCount, score) {
+			break
+		}
+	}
```

Comment:
This is needed b/c we're running Score Plugins in parallel; otherwise, we may be encountering either dirty read or racing issue.

https://github.com/kubernetes/kubernetes/blob/037751e7ad2cd18db5b4e2a20ba894314c522b15/pkg/scheduler/framework/v1alpha1/framework.go#L390-L391

Reply:
I feel this is not necessary, we can iterate over the nodes once in NormalizeScore to find the minCount. ```Score``` is not really meant to generate/modify shared state, it should only be concerned with scoring a node (i.e., should be stateless).

Reply:
I moved it into `NomalizedScore` in a foo loop.